Add Data2Vec

edugp opened the pull request:
# What does this PR do?

Add Data2Vec to transformers.

I started by cloning the RoBERTa model and fixing the conversion script for Data2Vec. We then added the audio model using some components from Wav2Vec2.

Conversion logs show identical forward passes for both text and audio:
```
max_absolute_diff = 0.0
Do both models output the same tensors? 🔥
```
Example usage for text:
```python
from transformers import RobertaTokenizer, Data2VecTextForSequenceClassification, Data2VecTextConfig
import torch

tokenizer = RobertaTokenizer.from_pretrained("roberta-large")
config = Data2VecTextConfig.from_pretrained("facebook/data2vec-text-base")
model = Data2VecTextForSequenceClassification.from_pretrained("facebook/data2vec-text-base", config=config)
# Fine-tune this model

inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
outputs = model(**inputs)

prediction_logits = outputs.logits
```
`data2vec-text-base` converted weights are [here](https://huggingface.co/facebook/data2vec-text-base), from [fairseq original weights](https://dl.fbaipublicfiles.com/fairseq/data2vec/nlp_base.pt)
`data2vec-audio-base` converted weights are [here](facebook/data2vec-audio-base) from [fairseq original weights](https://dl.fbaipublicfiles.com/fairseq/data2vec/audio_base_ls.pt) with no finetuning
`data2vec-audio-base-960h` converted weights are [here](https://huggingface.co/facebook/data2vec-audio-base-960h) from [fairseq original weights](https://dl.fbaipublicfiles.com/fairseq/data2vec/audio_base_ls_960h.pt) fine-tuned on 960 hours of Librispeech
NOTE: Data2Vec image model weights have not been released yet.
NOTE: The current implementation does not support pre-training yet, only fine-tuning on text and audio tasks.


Fixes # (issue)

## Who can review?

@patil-suraj @patrickvonplaten @mrm8488 @anton-l

Commit:
Add data2vec model cloned from roberta

Commit:
Add checkpoint conversion script

Commit:
Fix copies

Commit:
Fix copies

Commit:
Update docs

Commit:
Update docs

Commit:
Add checkpoint conversion script

HuggingFaceDocBuilder commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15507). All of your documentation changes will be reflected on that endpoint.

Commit:
Remove fairseq data2vec_text script and fix format

Commit:
Remove fairseq data2vec_text script and fix format

Commit:
Add comment on where to get data2vec_text.py

Commit:
Add comment on where to get data2vec_text.py

Commit:
Remove mock implementation cheat.py and fix style

Commit:
Remove mock implementation cheat.py and fix style

Commit:
Fix copies

Commit:
Fix copies

Commit:
Remove TF and Flax classes from init

Commit:
Remove TF and Flax classes from init

Commit:
Add back copy from fairseq data2vec_text.py and fix style

Commit:
Add back copy from fairseq data2vec_text.py and fix style

Commit:
Update model name in docs/source/index.mdx to be CamelCase

Commit:
Update model name in docs/source/index.mdx to be CamelCase

Commit:
Revert model name in table to lower-case to get check_table test to pass

Commit:
Revert model name in table to lower-case to get check_table test to pass

patrickvonplaten commented on src/transformers/__init__.py:
```diff
@@ -2928,6 +2943,17 @@
+        from .models.data2vec import (
+            DATA2VEC_PRETRAINED_MODEL_ARCHIVE_LIST,
+            Data2VecForCausalLM,
+            Data2VecForMaskedLM,
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/data2vec/__init__.py:
```diff
@@ -0,0 +1,60 @@
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

patrickvonplaten commented on src/transformers/models/data2vec/configuration_data2vec.py:
```diff
@@ -0,0 +1,73 @@
+}
+
+
+class Data2VecConfig(BertConfig):
```
```suggestion
class Data2VecConfig():
```

Could you please copy-paste all the code from `BertConfig` here? We are trying to not continue the abstraction from configs

edugp replied:
done!

patrickvonplaten commented on src/transformers/models/data2vec/convert_data2vec_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,185 @@
+# coding=utf-8
+# Copyright 2018 The HuggingFace Inc. team.
```
```suggestion
# Copyright 2022 The HuggingFace Inc. team.
```

patrickvonplaten commented on src/transformers/models/data2vec/convert_data2vec_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,185 @@
+from transformers.utils import logging
+
+# File copied from https://github.com/pytorch/fairseq/blob/main/examples/data2vec/models/data2vec_text.py
+from .data2vec_text import Data2VecTextModel
```
Very good idea to port a model! Guess we can delete this file shortly before merging

edugp replied:
I thought about just keeping the comment but not committing the file initially, but it was causing a CI test error, so I ended up adding it. I'll try removing it at the end and workaround the CI issue

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+# coding=utf-8
+# Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
```
```suggestion
# Copyright 2022 The HuggingFace Inc. team.
```

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+# coding=utf-8
+# Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+
+
+# Copied from transformers.models.roberta.modeling_roberta.RobertaEmbeddings with Roberta->Data2Vec
+class Data2VecEmbeddings(nn.Module):
```
Perfect! @patil-suraj - what do you think about the naming here? Is `Data2VecEmbeddingsForText` or `Data2VecTextEmbeddings` maybe better?

edugp replied:
I was thinking maybe `Data2VecForTextEmbeddings`, to match the model name `Data2VecForTextModel`. Or we can change the model name to `Data2VecTextModel`.

patrickvonplaten replied:
works for me as well - whatever you prefer :-)

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+    Same as BertEmbeddings with a tiny tweak for positional embeddings indexing.
+    """
+
+    # Copied from transformers.models.bert.modeling_bert.BertEmbeddings.__init__
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+                persistent=False,
+            )
+
+        # End copy
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+        return layer_output
+
+
+# Copied from transformers.models.bert.modeling_bert.BertEncoder with Bert->Data2Vec
```
Nice!

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+    DATA2VEC_START_DOCSTRING,
+)
+# Copied from transformers.models.roberta.modeling_roberta.RobertaModel with ROBERTA->DATA2VEC,Roberta->Data2Vec
+class Data2VecModel(Data2VecPreTrainedModel):
```
Here we need to change the name to something like `Data2VecForTextModel` since this is text specific only

edugp replied:
done!

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+@add_start_docstrings(
+    """data2vec Model with a `language modeling` head on top for CLM fine-tuning.""", DATA2VEC_START_DOCSTRING
+)
+# Copied from transformers.models.roberta.modeling_roberta.RobertaForCausalLM with ROBERTA->DATA2VEC,Roberta->Data2Vec,roberta->data2vec,roberta-base->data2vec
```
```suggestion
# Copied from transformers.models.roberta.modeling_roberta.RobertaForCausalLM with ROBERTA->DATA2VEC,Roberta->Data2Vec,roberta->data2vec,roberta-base->data2vec, RobertaModel->Data2VecForText
```

patrickvonplaten reviewed (commented):
This already looks great! Good job on porting the weights and making it work for Text. I think next we can take a stab at the Speech model weights. Here we should be able to fully re-utilize the `Data2VecEncoder` class and only add a new "front-end" for speech maybe called it `Data2VecAudioPreprocessor` or something like this

sgugger commented on docs/source/model_doc/data2vec.mdx:
```diff
@@ -0,0 +1,69 @@
+<!--Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
<!--Copyright 2022 The HuggingFace Team. All rights reserved.
```

sgugger commented on docs/source/model_doc/data2vec.mdx:
```diff
@@ -0,0 +1,69 @@
+specific language governing permissions and limitations under the License.
+-->
+
+# data2vec
```
```suggestion
# Data2Vec
```
I think the cased name is Data2Vec

sgugger commented on src/transformers/models/data2vec/configuration_data2vec.py:
```diff
@@ -0,0 +1,73 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" data2vec configuration"""
```
```suggestion
""" Data2Vec configuration"""
```

sgugger commented on src/transformers/models/data2vec/convert_data2vec_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,185 @@
+
+        # self attention
+        self_attn: BertSelfAttention = layer.attention.self
+        assert (
+            data2vec_layer.self_attn.k_proj.weight.data.shape
+            == data2vec_layer.self_attn.q_proj.weight.data.shape
+            == data2vec_layer.self_attn.v_proj.weight.data.shape
+            == torch.Size((config.hidden_size, config.hidden_size))
+        )
```
This should probably be split in several asserts, each with their error message?

edugp replied:
Thanks! I split this one and added error messages on these ones and all the other assertions

sgugger commented on src/transformers/models/data2vec/convert_data2vec_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,185 @@
+
+        # self-attention output
+        self_output: BertSelfOutput = layer.attention.output
+        assert self_output.dense.weight.shape == data2vec_layer.self_attn.out_proj.weight.shape
```
Same here, let's at least have a failure message.

sgugger commented on src/transformers/models/data2vec/convert_data2vec_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,185 @@
+
+        # intermediate
+        intermediate: BertIntermediate = layer.intermediate
+        assert intermediate.dense.weight.shape == data2vec_layer.fc1.weight.shape
```
Same here.

sgugger commented on src/transformers/models/data2vec/convert_data2vec_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,185 @@
+
+        # output
+        bert_output: BertOutput = layer.output
+        assert bert_output.dense.weight.shape == data2vec_layer.fc2.weight.shape
```
Same here.

sgugger commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch data2vec model."""
```
```suggestion
"""PyTorch Data2Vec model."""
```

sgugger commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+
+
+@add_start_docstrings(
+    """data2vec Model with a `language modeling` head on top for CLM fine-tuning.""", DATA2VEC_START_DOCSTRING
```
```suggestion
    """Data2Vec Model with a `language modeling` head on top for CLM fine-tuning.""", DATA2VEC_START_DOCSTRING
```

sgugger commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -0,0 +1,1577 @@
+
+@add_start_docstrings(
+    """
+    data2vec Model transformer with a sequence classification/regression head on top (a linear layer on top of the
```
```suggestion
    Data2Vec Model transformer with a sequence classification/regression head on top (a linear layer on top of the
```

sgugger commented on tests/test_modeling_data2vec.py:
```diff
@@ -0,0 +1,556 @@
+# coding=utf-8
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

sgugger reviewed (approved):
Very nice! Great beta-testing of the new add-model-like feature :-)

edugp committed:
Update src/transformers/models/data2vec/__init__.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/__init__.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/convert_data2vec_original_pytorch_checkpoint_to_pytorch.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/convert_data2vec_original_pytorch_checkpoint_to_pytorch.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update docs/source/model_doc/data2vec.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update docs/source/model_doc/data2vec.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/auto/configuration_auto.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/auto/configuration_auto.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/configuration_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/configuration_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update tests/test_modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update tests/test_modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/configuration_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/configuration_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Commit:
Update documentation

Commit:
Update documentation

Commit:
Merge branch 'add-data2vec-from-roberta' of https://github.com/edugp/transformers into add-data2vec-from-roberta

Commit:
Copy-paste Data2VecConfig from BertConfig

Commit:
Copy-paste Data2VecConfig from BertConfig

edugp commented on src/transformers/models/data2vec/configuration_data2vec.py:
```diff
@@ -0,0 +1,73 @@
+logger = logging.get_logger(__name__)
+
+DATA2VEC_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "data2vec": "https://huggingface.co/data2vec/resolve/main/config.json",
```
I can set it to `edugp/data2vec-nlp-base`, for now

Commit:
Update config checkpoint to point to edugp/data2vec-nlp-base. Fix style and repo-consistency

Commit:
Update config checkpoint to point to edugp/data2vec-nlp-base. Fix style and repo-consistency

Commit:
Update config special tokens to match RoBERTa

Commit:
Update config special tokens to match RoBERTa

Commit:
Split multiple assertions and add individual error messages

Commit:
Split multiple assertions and add individual error messages

Commit:
Rename Data2VecModel to Data2VecForTextModel

Commit:
Rename Data2VecModel to Data2VecForTextModel

Commit:
Add Data2Vec to _toctree.yml

Commit:
Add Data2Vec to _toctree.yml

Commit:
Rename Data2VecEmbeddings to Data2VecForTextEmbeddings

Commit:
Rename Data2VecEmbeddings to Data2VecForTextEmbeddings

Commit:
Add initial Data2VecForAudio model (unfinished). Only matching fairseq's implementation up to the feature encoder (before positional encoding).

Commit:
Add initial Data2VecForAudio model (unfinished). Only matching fairseq's implementation up to the feature encoder (before positional encoding).

edugp commented on src/transformers/models/data2vec/ema_module.py:
```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
+
+# TODO: Remove file
```
This was copied from fairseq's codebase to help in the conversion of model weights and should be removed before merging this PR

patrickvonplaten replied:
think we can delete it now

edugp replied:
I deleted all the fairseq files

edugp commented on src/transformers/models/data2vec/data2vec_text.py:
```diff
@@ -0,0 +1,474 @@
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# TODO: Remove file
```
This was copied from fairseq's codebase to help in the conversion of model weights and should be removed before merging this PR

edugp commented on src/transformers/models/data2vec/data2vec_audio.py:
```diff
@@ -0,0 +1,481 @@
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# TODO: Remove file
```
This was copied from fairseq's codebase to help in the conversion of model weights and should be removed before merging this PR

patrickvonplaten replied:
Think we can delete it now

edugp commented on src/transformers/models/data2vec/configuration_data2vec.py:
```diff
@@ -115,6 +115,22 @@ def __init__(
         position_embedding_type="absolute",
         use_cache=True,
         classifier_dropout=None,
+        # Here starts audio-only parameters
```
We can refactor this to make a clear distinction between common parameters and modality-specific parameters

patrickvonplaten replied:
Good for now!

edugp commented on src/transformers/models/data2vec/convert_data2vec_audio_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,246 @@
+# coding=utf-8
```
To run the checkpoint conversion script:
```
python -m transformers.models.data2vec.convert_data2vec_audio_original_pytorch_checkpoint_to_pytorch --checkpoint_path <your_path_to>/audio_base_ls.pt --pytorch_dump_folder_path <your_output_path> --not_finetuned
```
`audio_base_ls.pt` can be downloaded from [here](https://dl.fbaipublicfiles.com/fairseq/data2vec/audio_base_ls.pt).  
You will see two print statements showing that the forward pass matches up until the addition of convolutional positional embeddings.

edugp commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -45,6 +45,13 @@
     find_pruneable_heads_and_indices,
     prune_linear_layer,
 )
+from ...models.wav2vec2.modeling_wav2vec2 import (
```
Do we want to copy these over, or importing them from `modeling_wav2vec2` is fine? @patrickvonplaten

patrickvonplaten replied:
We should copy all of them over with Copied from statements ;-)

edugp commented on src/transformers/models/data2vec/modeling_data2vec.py:
```diff
@@ -684,7 +691,83 @@ def update_keys_to_ignore(self, config, del_keys_to_ignore):
+
+        self.adapter = Wav2Vec2Adapter(config) if config.add_adapter else None
+
+        self.encoder = Data2VecEncoder(config)
```
Note we are re-using the same encoder as `Data2VecForText`.

patrickvonplaten committed:
finish audio model

patrickvonplaten committed:
finish audio model

patrickvonplaten committed:
finish audio file

patrickvonplaten committed:
finish audio file

Commit:
Update names and fix style, quality and repo consistency

Commit:
Update names and fix style, quality and repo consistency

Commit:
Remove Data2VecAudioForPretraining. Add tests for Data2VecAudio, mimicking the Wav2Vec2 test suite. Fix bias initilization in positional conv layers. Move back configurations for audio and text to separate files.

Commit:
Remove Data2VecAudioForPretraining. Add tests for Data2VecAudio, mimicking the Wav2Vec2 test suite. Fix bias initilization in positional conv layers. Move back configurations for audio and text to separate files.

patrickvonplaten committed:
Merge branch 'master' of https://github.com/huggingface/transformers into add-data2vec-from-roberta

patrickvonplaten committed:
add inputs to logits to data2vec'

patrickvonplaten committed:
add inputs to logits to data2vec'

patrickvonplaten committed:
Merge branch 'add-data2vec-from-roberta' of https://github.com/edugp/transformers into add-data2vec-from-roberta

patrickvonplaten committed:
correct autio models

patrickvonplaten committed:
correct autio models

patrickvonplaten committed:
correct config auto

patrickvonplaten committed:
correct config auto

patrickvonplaten committed:
correct tok auto

patrickvonplaten committed:
correct tok auto

patrickvonplaten commented on utils/tests_fetcher.py:
```diff
@@ -345,6 +345,7 @@ def module_to_test_file(module_fname):
     "tests/sagemaker/test_single_node_gpu.py",  # SageMaker test
     "tests/sagemaker/test_multi_node_model_parallel.py",  # SageMaker test
     "tests/sagemaker/test_multi_node_data_parallel.py",  # SageMaker test
+    "tests/test_modeling_data2vec.py",  # TODO:Remove redundant file
```
```suggestion
```

patrickvonplaten committed:
Update utils/tests_fetcher.py

patrickvonplaten committed:
Update utils/tests_fetcher.py

patrickvonplaten committed:
delete unnecessary files

patrickvonplaten committed:
delete unnecessary files

patrickvonplaten committed:
delete unnecessary files

patrickvonplaten committed:
delete unnecessary files

patrickvonplaten committed:
Merge branch 'add-data2vec-from-roberta' of https://github.com/edugp/transformers into add-data2vec-from-roberta

patrickvonplaten committed:
further renaming

patrickvonplaten committed:
further renaming

patrickvonplaten committed:
make all tests pass

patrickvonplaten committed:
make all tests pass

patrickvonplaten committed:
finish

patrickvonplaten committed:
finish

patrickvonplaten committed:
remove useless test file

patrickvonplaten committed:
remove useless test file

patrickvonplaten commented on docs/source/model_doc/data2vec.mdx:
```diff
@@ -0,0 +1,110 @@
+
+## Overview
+
+The Data2Vec model was proposed in [data2vec: A General Framework for Self-supervised Learning in Speech, Vision and Language](https://scontent-sjc3-1.xx.fbcdn.net/v/t39.8562-6/271974914_483120576492438_4239522333319653600_n.pdf?_nc_cat=107&ccb=1-5&_nc_sid=ae5e01&_nc_ohc=7huShTb_QZIAX-N7SYx&_nc_ht=scontent-sjc3-1.xx&oh=00_AT_lXXL69mjqmdVWbaLh4Ro6DY17aFeO5vA9I-mIpyNieg&oe=6205C411) by Alexei Baevs ...
```
The link looks a bit weird - can we change it?

patrickvonplaten replied:
there is surely an arxiv link no?

edugp replied:
Indeed! I'll change it to this [link](https://arxiv.org/pdf/2202.03555.pdf)

patrickvonplaten commented on docs/source/model_doc/data2vec.mdx:
```diff
@@ -0,0 +1,110 @@
+
+Tips:
+
+- This implementation has a shared encoder for all different modalities and different pre-processors for each modality.
```
Think this needs to be updated. Maybe we can just write that both Data2VecAudio and Data2VecText have been pretrained using the same method

patrickvonplaten commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -321,10 +325,14 @@
 
-SPECIAL_MODEL_TYPE_TO_MODULE_NAME = OrderedDict([("openai-gpt", "openai")])
+SPECIAL_MODEL_TYPE_TO_MODULE_NAME = OrderedDict(
+    [("openai-gpt", "openai"), ("data2vec-audio", "data2vec"), ("data2vec-text", "data2vec")]
```
Is this ok @sgugger ? Added it like this now

patrickvonplaten commented on src/transformers/models/data2vec/configuration_data2vec_text.py:
```diff
@@ -0,0 +1,147 @@
+    This is the configuration class to store the configuration of a [`Data2VecTextModel`] and [`Data2VecTextModel`]. It
+    is used to instantiate a Data2VecText model according to the specified arguments, defining the model architecture.
+    Instantiating a configuration with the defaults will yield a similar configuration to that of the Data2VecText
+    [edugp/data2vec-nlp-base](https://huggingface.co/edugp/data2vec-nlp-base) architecture.
```
actually can we update this checkpoint to https://huggingface.co/facebook/data2vec-text-base ?

edugp replied:
done!

patrickvonplaten commented on src/transformers/models/data2vec/configuration_data2vec_text.py:
````diff
@@ -0,0 +1,147 @@
+    ```python
+    >>> from transformers import Data2VecTextModel, Data2VecTextConfig
+
+    >>> # Initializing a Data2VecText edugp/data2vec-nlp-base style configuration
````
```suggestion
    >>> # Initializing a Data2VecText [edugp/data2vec-nlp-base](https://huggingface.co/facebook/data2vec-text-base) style configuration
```

patrickvonplaten commented on src/transformers/models/data2vec/configuration_data2vec_text.py:
```diff
@@ -0,0 +1,147 @@
+    >>> # Initializing a Data2VecText edugp/data2vec-nlp-base style configuration
+    >>> configuration = Data2VecTextConfig()
+
+    >>> # Initializing a model from the edugp/data2vec-nlp-base style configuration
```
```suggestion
    >>> # Initializing a model from the https://huggingface.co/facebook/data2vec-text-base style configuration
```

patrickvonplaten commented on src/transformers/models/data2vec/configuration_data2vec_text.py:
````diff
@@ -0,0 +1,147 @@
+    ```python
+    >>> from transformers import Data2VecTextModel, Data2VecTextConfig
+
+    >>> # Initializing a Data2VecText edugp/data2vec-nlp-base style configuration
````
same here -> https://huggingface.co/facebook/data2vec-text-base

edugp replied:
done!

patrickvonplaten commented on src/transformers/models/data2vec/configuration_data2vec_text.py:
```diff
@@ -0,0 +1,147 @@
+    >>> # Initializing a Data2VecText edugp/data2vec-nlp-base style configuration
+    >>> configuration = Data2VecTextConfig()
+
+    >>> # Initializing a model from the edugp/data2vec-nlp-base style configuration
```
and here: https://huggingface.co/facebook/data2vec-text-base

edugp replied:
done!

patrickvonplaten commented on src/transformers/models/data2vec/data2vec_text.py:
```diff
@@ -0,0 +1,474 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
```
think we can delete it now as well

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+    attentions: Optional[Tuple[torch.FloatTensor]] = None
+
+
+def _compute_mask_indices(
```
can we add a # Copied from statement here from Wav2Vec2?

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+
+
+@dataclass
+class XVectorOutput(ModelOutput):
```
Think we can also add a # Copied from statement here (from Wav2Vec2)

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        return hidden_states
+
+
+class Data2VecAudioPadLayer(nn.Module):
```
Copied from statement?

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        return hidden_states
+
+
+class Data2VecAudioFeatureEncoder(nn.Module):
```
Can we add a Copied from statement here?

edugp replied:
I removed a few of the `Copy from` statements because `make fixup` complaining that the classes didn't look identical and that I had to run `make-copies` to replace it with the original one, something along those lines, which was okay in some cases but in others the classes needed actual modifications.
I will add `Copy from` statements to all for now and we can figure out in the end which ones need to be removed (if any)

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        return hidden_states
+
+
+class Data2VecAudioFeatureProjection(nn.Module):
```
Copied from? From Wav2Vec2

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        return hidden_states
+
+
+class Data2VecAudioEncoderLayer(nn.Module):
```
Also Copied from here from Wav2Vec2 I tihnk

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        return attn_output, attn_weights_reshaped, past_key_value
+
+
+class Data2VecAudioFeedForward(nn.Module):
```
Copied from from Wav2Vec2 no?

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        return outputs
+
+
+class Data2VecAudioEncoder(nn.Module):
```
Here as well we can add a Copied from I think

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        )
+
+
+class Data2VecAudioAdapter(nn.Module):
```
Copied from from Wav2Vec2

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        return hidden_states
+
+
+class Data2VecAudioAdapterLayer(nn.Module):
```
Also a Copied from here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+                k = math.sqrt(module.groups / (module.in_channels * module.kernel_size[0]))
+                nn.init.uniform_(module.bias, a=-k, b=k)
+
+    def _get_feat_extract_output_lengths(
```
Think we can add a "Copied from ..." here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+
+        return input_lengths
+
+    def _get_feature_vector_attention_mask(
```
Think we can add a "Copied from ..." here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+
+
+DATA2VEC_AUDIO_START_DOCSTRING = r"""
+    Data2VecAudio was proposed in [wav2vec 2.0: A Framework for Self-Supervised Learning of Speech
```
This docstring needs to be updated

patrickvonplaten replied:
with the correct paper etc...

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+    """Data2VecAudio Model with a `language modeling` head on top for Connectionist Temporal Classification (CTC).""",
+    DATA2VEC_AUDIO_START_DOCSTRING,
+)
+class Data2VecAudioForCTC(Data2VecAudioPreTrainedModel):
```
Think we can also add a "Copied from statement here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+    """,
+    DATA2VEC_AUDIO_START_DOCSTRING,
+)
+class Data2VecAudioForSequenceClassification(Data2VecAudioPreTrainedModel):
```
Can also add a Copied from here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+    """,
+    DATA2VEC_AUDIO_START_DOCSTRING,
+)
+class Data2VecAudioForAudioFrameClassification(Data2VecAudioPreTrainedModel):
```
Can also add a Copied from here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        )
+
+
+class AMSoftmaxLoss(nn.Module):
```
Can also add a Copied from here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+        return loss
+
+
+class TDNNLayer(nn.Module):
```
Can also add a Copied from here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_audio.py:
```diff
@@ -0,0 +1,1513 @@
+    """,
+    DATA2VEC_AUDIO_START_DOCSTRING,
+)
+class Data2VecAudioForXVector(Data2VecAudioPreTrainedModel):
```
Can also add a Copied from here

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_text.py:
```diff
@@ -0,0 +1,1578 @@
+_TOKENIZER_FOR_DOC = "RobertaTokenizer"
+
+# General docstring
+_CONFIG_FOR_DOC = "Wav2Vec2Config"
```
This looks wrong - do we even need this here?

edugp replied:
great catch, the line after looks wrong too

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_text.py:
```diff
@@ -0,0 +1,1578 @@
+_PROCESSOR_FOR_DOC = "Wav2Vec2Processor"
+
+# CTC docstring
+_CTC_EXPECTED_OUTPUT = "'MISTER QUILTER IS THE APOSTLE OF THE MIDDLE CLASSES AND WE ARE GLAD TO WELCOME HIS GOSPEL'"
```
Same here think we can delete this and the following line

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_text.py:
```diff
@@ -0,0 +1,1578 @@
+        return position_ids.unsqueeze(0).expand(input_shape)
+
+
+class Data2VecTextSelfAttention(nn.Module):
```
Can't we add a "Copied from ..." here from Roberta?

patrickvonplaten commented on src/transformers/models/data2vec/modeling_data2vec_text.py:
```diff
@@ -0,0 +1,1578 @@
+        >>> from transformers import Data2VecTextTokenizer, Data2VecTextForCausalLM, Data2VecTextConfig
+        >>> import torch
+
+        >>> tokenizer = Data2VecTextTokenizer.from_pretrained("data2vec-base")
```
and everywhere below as well

patrickvonplaten commented on tests/test_modeling_common.py:
```diff
@@ -1911,6 +1914,7 @@ def test_load_with_mismatched_shapes(self):
         config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
 
         for model_class in self.all_model_classes:
+            print(model_class)
```
```suggestion
```

patrickvonplaten committed:
Update tests/test_modeling_common.py

patrickvonplaten committed:
Update tests/test_modeling_common.py

patrickvonplaten commented on utils/check_repo.py:
```diff
@@ -168,8 +169,19 @@
     "VisualBertForMultipleChoice",
     "TFWav2Vec2ForCTC",
     "TFHubertForCTC",
+    "Data2VecAudioModel",  # TODO: Fix
```
This can be removed
```suggestion
```

patrickvonplaten commented on utils/check_repo.py:
```diff
@@ -168,8 +169,19 @@
 
+# Update this list for models that have multiple model types for the same
+# model doc
+MODEL_TYPE_TO_DOC_MAPPING = OrderedDict(
```
@sgugger - not really sure if this is ok here, but didn't really manage to fix the utils/doc_test otherwise

patrickvonplaten replied:
@osanseviero @julien-c - can we make this exception here for the model_type to doc link? The reason is that Data2Vec is one paper, one pretraining approach, but the architectures for audio, text and vision are so different that we need different modeling files, e.g.:
- https://huggingface.co/models?other=data2vec-text
- https://huggingface.co/models?other=data2vec-audio 

should both map to https://huggingface.co/docs/transformers/model_doc/data2vec

Wolud this be easy to implement or problematic?

julien-c replied:
we could maybe just redirect on the doc side no?

patrickvonplaten replied:
In a sense that we create `https://huggingface.co/docs/transformers/model_doc/data2vec-text` and `https://huggingface.co/docs/transformers/model_doc/data2vec-audio` which redirect to `https://huggingface.co/docs/transformers/model_doc/data2vec`? 

@sgugger - any ideas on how to move forward here?

patrickvonplaten replied:
BTW, I'm not a huge fan in general that we make the Transformers library more and more dependent on the HF Hub. I don't think design decisions of a fully open-sourced library like Transformers should be driven in any way by higher level tools that make use of the library, like the HF Hub. 

Think another small example of this is that the widgets on the Hub are tied to the naming of classes in Transformers.

So far these are very minor dependencies and can easily be respected from the Transformers side, but they do limit Transformer's design a bit unnecessarily IMO and slow down open-source contributions (*e.g.* how would an open-source contributor know the points mentioned above?)

Keen to hear your opinion on this, also cc @LysandreJik

sgugger replied:
I had several ideas of things to test, but it looks like you did not wait for any of them, so I guess this is going to be this way.

julien-c replied:
@sgugger maybe we can find with Mishig a metadata or JS tag in an .mdx file that would mean `redirect to this other doc page`? cc @Pierrci also

mrm8488 replied:
I think so

patrickvonplaten reviewed (approved):
Amazing work @edugp ! Really cool to see how you managed to quickly debug the Fairseq model and made the models work!

In general the PR looks like it's in its final steps. Think it would be important to add a bunch of `"Copied from ..."` statements (from modeling_wav2vec2.py) to make the code more robust, but the main parts look very nice now!

Also, I've locally checked that all the slow tests run correctly and uploaded the official checkpoints under the Facebook org here:
https://huggingface.co/models?other=data2vec-text
https://huggingface.co/models?other=data2vec-audio

@sgugger - I did a pretty in-depth review here. It would be very nice if you could take a quick look at the changes we've done to make the doc tests and auto config work

patrickvonplaten commented:
Ah the new test design just got merged [here](https://github.com/huggingface/transformers/pull/15725) :sweat_drops: 

@edugp - I'm afraid you'll have to merge current master into your branch here (or do `git rebase`) and resolve the conflicts before we can merge this one. Would be amazing if you could give it a stab - should be pretty easy I think, just create a test dir called for data2vec and move the files there :-)

edugp committed:
Update utils/check_repo.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update utils/check_repo.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec_text.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Update src/transformers/models/data2vec/modeling_data2vec_text.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

edugp committed:
Merge branch 'huggingface:master' into master

Commit:
Fix conflict

Commit:
Move data2vec tests to new structure

Commit:
Fix test imports for text tests

Commit:
Remove fairseq files

Commit:
Change paper link to arxiv

Commit:
Modify Data2Vec documentation to reflect that the encoder is not shared across the audio and text models in the current implementation.

Commit:
Update text model checkpoint to be facebook/data2vec-text-base

Commit:
Add 'Copy from' statements and update paper links and docs

edugp commented:
Thanks @patrickvonplaten, I synced the latest master with this branch and moved the tests to the new directory structure (and verified that they still pass). I also added all `Copied from` statements and  addressed the rest of comments. Thanks for the thorough review!

patrickvonplaten committed:
Merge branch 'master' of https://github.com/huggingface/transformers into add-data2vec-from-roberta

patrickvonplaten committed:
fix copy from statements

patrickvonplaten committed:
improve copied from

patrickvonplaten committed:
correct more copied from statements

patrickvonplaten committed:
finish copied from stuff

patrickvonplaten committed:
make style

patrickvonplaten commented:
Fixed the red CI @edugp - should be good for merge now.

@sgugger - I'm not 100% whether it's ok what I've done here: https://github.com/huggingface/transformers/pull/15507/files#r813404423 . Given that we have one `data2vec` folder and one `data2vec` paper I think it's also better to have one doc page no? If you're ok with this change in the `utils/...` I think we can merge.

patrickvonplaten committed:
add model to README

patrickvonplaten committed:
add to master

wise-east commented:
> # What does this PR do?
> Add Data2Vec to transformers.
> 
> I started by cloning the RoBERTa model and fixing the conversion script for Data2Vec. We then added the audio model using some components from Wav2Vec2.
> 
> Conversion logs show identical forward passes for both text and audio:
> 
> ```
> max_absolute_diff = 0.0
> Do both models output the same tensors? 🔥
> ```
> 
> Example usage for text:
> 
> ```python
> from transformers import RobertaTokenizer, Data2VecTextForSequenceClassification, Data2VecTextConfig
> import torch
> 
> tokenizer = RobertaTokenizer.from_pretrained("roberta-large")
> config = Data2VecTextConfig.from_pretrained("facebook/data2vec-text-base")
> model = Data2VecTextForSequenceClassification.from_pretrained("facebook/data2vec-text-base", config=config)
> # Fine-tune this model
> 
> inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
> outputs = model(**inputs)
> 
> prediction_logits = outputs.logits
> ```
> 
> `data2vec-text-base` converted weights are [here](https://huggingface.co/facebook/data2vec-text-base), from [fairseq original weights](https://dl.fbaipublicfiles.com/fairseq/data2vec/nlp_base.pt) `data2vec-audio-base` converted weights are [here](facebook/data2vec-audio-base) from [fairseq original weights](https://dl.fbaipublicfiles.com/fairseq/data2vec/audio_base_ls.pt) with no finetuning `data2vec-audio-base-960h` converted weights are [here](https://huggingface.co/facebook/data2vec-audio-base-960h) from [fairseq original weights](https://dl.fbaipublicfiles.com/fairseq/data2vec/audio_base_ls_960h.pt) fine-tuned on 960 hours of Librispeech NOTE: Data2Vec image model weights have not been released yet.
> 
> Fixes # (issue)
> 
> ## Who can review?
> @patil-suraj @patrickvonplaten @mrm8488 @anton-l

Is this supposed to work? I'm getting `ImportError: cannot import name 'Data2VecTextConfig' from 'transformers' (/data/home/justincho/Data2Vec_lightning/transformers/src/transformers/__init__.py)` after cloning this branch and using it to install the transformers library : `https://github.com/edugp/transformers/tree/add-data2vec-from-roberta`

wise-east commented:
Does this update include changes necessary for training using the data2vec objective? I'm curious because I don't see an EMA implementation anywhere.

edugp commented:
> Does this update include changes necessary for training using the data2vec objective? I'm curious because I don't see an EMA implementation anywhere.

@wise-east it does not support pre-training yet, unfortunately! Only fine-tuning on text and audio tasks. I will update the PR description to reflect that.

patrickvonplaten commented:
@sgugger @osanseviero @julien-c - merging this now. Sorry, I'm not 100% sure yet how to handle the HF Hub <=> Transformers doc page linking for this model, but I think we can solve this in a follow-up PR.

patrickvonplaten commented:
Great job @edugp :partying_face:

mrm8488 commented:
Enorme, @edugp !!!
